### PR TITLE
Flatpak Port

### PR DESCRIPTION
**Pull request by jo5huajohn**
I'm not sure if this is in a state to be merged, but I figured it would be good if other people with more experience with flatpaks took a look at this and suggested best practices/improvements that I could make.

### CODE REVIEW

**Comment by vrement** [comment]
Oh nice! The first "RC" is here

**Comment by berarma** [comment]
I know nothing about Flatpak packaging.  It doesn't affect normal operation , I can merge and let users try it.

Apart from merging, I guess there's some more work to do to create the actual package and publish it. Are you doing too?

Instead of patching the meson build file, wouldn't be easier to use the command line arguments to change the udev path?

**Comment by berarma** [comment]
There might be some important tips [here](https://github.com/berarma/oversteer/issues/139#issuecomment-1538503438). I would need help implementing them correctly for Flatpak though.

**Comment by jo5huajohn** [comment]
I've removed the patch in my local branch and added the config in the metadata file in my local workspace.

**Comment by jo5huajohn** [comment]
As for creating the actual package and supporting it, I'd like to do it, but it will take a while since I'm still learning what goes into creating a flatpak. I'm also looking into how to support XDG portals in the application. From what I understand, oversteer only needs to integrate the file chooser portal to select profiles and test CSVs.

**Comment by jo5huajohn** [comment]
I've implemented the xdg file chooser portal and updated the appdata file, but I'm not so sure about the udev rules since as I understand it, it violates flatpak rules.

**Comment by jo5huajohn** [comment]
Additionally, I would like to rename the app-id to io.github.berarma.oversteer as per the flathub app id requirements:
>The Application ID should be a real URL of a domain that the app author has control over or where their app is hosted.
>
>Ignoring this will lead to problems down the line, such as not being able to verify the app and receiving payments. It also decides, which verification methods will be available. For e.g. using io.github.flathub.TestApp would only allow for Github or Website verification.

**Comment by jo5huajohn** [comment]
This change is to implement the FileChooser portal, and introduce file filters for tests and profiles.  [FileChooserNative](https://docs.gtk.org/gtk3/class.FileChooserNative.html) uses FileChooserDialog in the backend, but uses the FileChooser portal if it detects that it is being run in a flatpak environment.

**Comment by berarma** [comment]
> This change is to implement the FileChooser portal, and introduce file filters for tests and profiles. [FileChooserNative](https://docs.gtk.org/gtk3/class.FileChooserNative.html) uses FileChooserDialog in the backend, but uses the FileChooser portal if it detects that it is being run in a flatpak environment.

Why are they duplicated? Once in the source code and the same changes again in portals.diff.

**Comment by jo5huajohn** [comment]
Oops. That shouldn't have been there. Removed.

**Comment by jo5huajohn** [comment]
> Additionally, I would like to rename the app-id to io.github.berarma.oversteer as per the flathub app id requirements:
> 
> > The Application ID should be a real URL of a domain that the app author has control over or where their app is hosted.
> > Ignoring this will lead to problems down the line, such as not being able to verify the app and receiving payments. It also decides, which verification methods will be available. For e.g. using io.github.flathub.TestApp would only allow for Github or Website verification.

Do you have any thoughts on this? Additionally, if the flatpak is good to go, the flatpak directory should be submitted to the flathub repo. I can add a commit to remove the flatpak directory and submit it to the flathub repo.

**Comment by berarma** [comment]
> > Additionally, I would like to rename the app-id to io.github.berarma.oversteer as per the flathub app id requirements:
> > > The Application ID should be a real URL of a domain that the app author has control over or where their app is hosted.
> > > Ignoring this will lead to problems down the line, such as not being able to verify the app and receiving payments. It also decides, which verification methods will be available. For e.g. using io.github.flathub.TestApp would only allow for Github or Website verification.
> 
> Do you have any thoughts on this? Additionally, if the flatpak is good to go, the flatpak directory should be submitted to the flathub repo. I can add a commit to remove the flatpak directory and submit it to the flathub repo.

I've enabled berarma.gihub.io. Is that enough?

**Comment by jo5huajohn** [comment]
It isn't necessary, but yeah. I've made the following changes:
- Removed the flatpak directory since this needs to be submitted to the flathub repo.
- Included an image in data/ and referenced this in the README and the appdata file (for flathub images).
- Renamed the application in all appropriate files.

Once this commit is merged, I'll run final tests defined by the flathub guidelines and create a new PR there.

**Comment by jo5huajohn** [comment]
A flatpak for Oversteer is now live on [flathub](https://flathub.org/apps/io.github.berarma.Oversteer).

**Comment by berarma** [comment]
> A flatpak for Oversteer is now live on [flathub](https://flathub.org/apps/io.github.berarma.Oversteer).

Shouldn't I make a release with the Flatpak changes?

I've tried it but it doesn't read settings nor let me change them. It has no permissions.

**Comment by jo5huajohn** [comment]
Ideally, yes. But I've linked it to the latest commit for now. I'll start following the releases once 0.8.3 is out.

That's odd. I'm assuming you're running the driver and have the udev rules installed. Do you have xdg-desktop-portal-gnome on your system? You can check by going to the directory `/usr/share/xdg-desktop-portal` and looking for `gnome-portals.conf`.

You can also try `flatpak override --reset io.github.berarma.Oversteer`. This should reset any global permission overrides on the application. Some users have also reported a reboot working.

My suspect is the application can't access the files created by the driver due to flatpak filesystem permissions on your system.

You can also use Flatseal to check if any permissions are denied.

**Comment by berarma** [comment]
I uninstalled my natively installed version and it removed the udev rules. Now that I've restored them it's working. I'll add a note in the Readme. I'll create a new release shortly so Flatpak users are on a known release version.

Thanks for your work on this! :clap:

**Comment by jo5huajohn** [comment]
Awesome. Happy to help!

**Comment by vrement** [comment]
Just noticed this after a longer time and i see we now have a flatpak! Great work all involved! Viva la easy effortless Linux installations :+1:

**Comment by vandea05** [comment]
A note from a Linux neophyte- the readme for the Flatpak is a little sparse and I found the documentation pretty hard to follow regarding the udev rules. I'd propose updating the Flatpak readme to make the udev requirement a little more prominent and perhaps include a little more information on where to find the rules in the repo and where to put them. It probably seems really obvious from inside the ecosystem but it's quite daunting from the outside!

**Comment by jo5huajohn** [comment]
@vandea05 I'll have the README updated with the information in the next update.
